src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts

```diff
@@ -1389,11 +1389,60 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 	}
 
 	private executeTask(task: Task, resolver: ITaskResolver): Promise<ITaskSummary> {
-		return ProblemMatcherRegistry.onReady().then(() => {
-			return this.editorService.saveAll().then(() => { // make sure all dirty editors are saved
-				let executeResult = this.getTaskSystem().run(task, resolver);
-				return this.handleExecuteResult(executeResult);
+		enum saveBeforeRunConfigOptions {
+			Always = 'always',
+			Never = 'never',
+			Prompt = 'prompt'
+		}
+
+		const saveBeforeRunTaskConfig: saveBeforeRunConfigOptions = this.configurationService.getValue('task.saveBeforeRun');
+
+		const execTask = async (task: Task, resolver: ITaskResolver): Promise<ITaskSummary> => {
+			let executeResult = this.getTaskSystem().run(task, resolver);
+			return this.handleExecuteResult(executeResult);
+		};
+
+		const saveAllEditorsAndExecTask = async (task: Task, resolver: ITaskResolver): Promise<ITaskSummary> => {
+			return this.editorService.saveAll().then(() => {
+				return execTask(task, resolver);
 			});
+		};
+
+		const promptAsk = (task: Task, resolver: ITaskResolver): Promise<ITaskSummary> => {
+			const message = nls.localize('TaskSystem.saveBeforeRun.prompt.title', 'Save all editors and run task?');
+
+			let result: Promise<ITaskSummary> = new Promise<ITaskSummary>((resolve, reject) => {
+				resolve();
+			});
+
+			this.notificationService.prompt(Severity.Warning, message,
```

Comment:
Instead of showing a notification, which is ignoreable, we should show a dialog with `IDialogService`. Since recieving this choice is opt in with a setting, we should make sure the choice is not ignorreable.